ADBDEV-271

darthunix opened the pull request:
Fix parser algorithm for setting distribution key column numbers
in gp_distribution_policy relation for inherited tables in GP5X
(GP6X is ok).
Query planes with ORCA caused segmentation faults because of
out of range column numbers and legacy optimizer simply
returned error before current patch. For example:
```
create table ta (a int) distributed randomly;
create table tb (a int, b int) inherits (ta) distributed by (b);

set optimizer=on;
insert into tb values(0, 0);
-- Segmentation fault

set optimizer=off;
insert into tb values(0, 0);
-- ERROR: no tlist entry for key 3 (cdbmutate.c:1484)

select attrnums from gp_distribution_policy where localoid = 'tb'::regclass;
 attrnums
----------
{3}
(1 row)
```

Also a check for setting non-hashable distribution key from a
parent table in an inherited one didn't work.
```
create table tc (a point) distributed randomly;
create table td (b int) inherits (tc) distributed by (a);

select * from td;
ERROR:  could not find mergejoinable = operator for type 600 (pathkeys.c:1174)
```

darthunix committed:
Fix parser algorithm for setting distribution key column numbers
in gp_distribution_policy relation for inherited tables in GP5X
(GP6X is ok).
Query planes with ORCA caused segmentation faults because of
out of range column numbers and legacy optimizer simply
returned error before current patch. For example:

create table ta (a int) distributed randomly;
create table tb (a int, b int) inherits (ta) distributed by (b);

set optimizer=on;
insert into tb values(0, 0);
-- Segmentation fault

set optimizer=off;
insert into tb values(0, 0);
-- ERROR: no tlist entry for key 3 (cdbmutate.c:1484)

select attrnums from gp_distribution_policy where localoid = 'tb'::regclass;
 attrnums
----------
{3}
(1 row)

Also a check for setting non-hashable distribution key from a
parent table in an inherited one didn't work.

create table tc (a point) distributed randomly;
create table td (b int) inherits (tc) distributed by (a);

select * from td;
ERROR:  could not find mergejoinable = operator for type 600 (pathkeys.c:1174)

dursegov reviewed (commented):
It looks suspicious to me that table 'points3' is distributed by a, could you check the case with multiple inherited tables?

`postgres=# CREATE TABLE points1 (a int) distributed randomly;
CREATE TABLE
postgres=# CREATE TABLE points2 (a int, b int) distributed randomly;
CREATE TABLE
postgres=# CREATE TABLE points3 (a int, b int, c int) INHERITS (points1, points2) distributed by (b);
NOTICE:  merging multiple inherited definitions of column "a"
NOTICE:  merging column "a" with inherited definition
NOTICE:  merging column "b" with inherited definition
CREATE TABLE
postgres=# \d points3
    Table "public.points3"
 Column |  Type   | Modifiers 
--------+---------+-----------
 a      | integer | 
 b      | integer | 
 c      | integer | 
Inherits: points1,
          points2
Distributed by: (a)

postgres=# select * from gp_distribution_policy;
 localoid | attrnums 
----------+----------
    16441 | 
    16444 | 
    16447 | {1}
(3 rows)
`

leskin-in commented:
Please [run `pgindent`](https://github.com/greenplum-db/gpdb/tree/master/src/tools/pgindent) on code changes.

leskin-in commented:
Also note that some regression tests fail on my machine (Ubuntu 19.04). However, I suppose they are not connected to this PR.

In addition, the 5X integration test pipeline (unfortunately it is not open-source) is currently partially broken. This may have no influence on PR merge time, however.

darthunix committed:
Remove a dirty hack, updating distribution policy in
create table statement (seems to be a workaround for a
wrong algorithm in parser).

darthunix commented:
@dursegov your bug seems to be fixed. But I still need to run linters on this code (as @leskin-in has mentioned about pgident)

darthunix committed:
Fix code style for a changed code block with pgindent

darthunix commented:
I have fixed a code style for a changed code block with pgindent (though nobody cares about pgindend in 5X_STABLE). Seems all is ready for merging PR, yes?
